src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,962 @@
+			// TODO: Rx interrupt for CPU interface
+		}
+	}
+}
+
+void tandberg_tdv2100_disp_logic_device::uart_tx(int state)
+{
+	if(state)
+	{
+		bool tty_mode = m_dsw_u39->read()&0x010;
+		if(tty_mode)
+		{
+			if(tx_data_pending_kbd)
+			{
+				m_uart->transmit(tx_data_kbd);
+				tx_data_pending_kbd = false;
+			}
+		}
+		else
+		{
+			// TODO: Tx interrupt for CPU interface
+		}
+	}
+}
+
+INPUT_CHANGED_MEMBER(tandberg_tdv2100_disp_logic_device::uart_changed)
+{
+	set_uart_state_from_switches();
+}
+
+void tandberg_tdv2100_disp_logic_device::set_uart_state_from_switches()
+{
+	int baud_setting = m_sw_rs232_baud->read();
+	int u46_divisor = (baud_setting == 0x0) ? 15 : 11;
+	int u22_divisor = 256>>baud_setting;
+	m_uart_clock->set_unscaled_clock(DOT_CLOCK_3 / (u46_divisor * u22_divisor));
+	m_uart->write_cs(true);
+	m_uart->write_nb1(m_dsw_u39->read()&0x200);
+	m_uart->write_nb2(true);
+	m_uart->write_np(m_sw_rs232_settings->read()&0x01);
+	m_uart->write_eps(m_sw_rs232_settings->read()&0x02);
+	m_uart->write_tsb(m_sw_rs232_settings->read()&0x04);
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Strobes for Keyboard
+//
+
+void tandberg_tdv2100_disp_logic_device::process_keyboard_char(uint8_t key)
+{
+	if(key&0x80)
+	{
+		tx_data_kbd = key&0x7F;
+		tx_data_pending_kbd = true;
+		bool tty_mode = m_dsw_u39->read()&0x010;
+		if(tty_mode)
+		{
+			if(m_uart->tbmt_r())
+			{
+				uart_tx(1);
+			}
+		}
+		else
+		{
+			// TODO: Keyboard interrupt for CPU interface
+		}
+	}
+}
+
+void tandberg_tdv2100_disp_logic_device::w_break(int state)
+{
+	break_key_held = state;
+}
+
+void tandberg_tdv2100_disp_logic_device::w_cleark(int state)
+{
+	if(state)
+	{
+		m_write_errorl_cb(0);
+		bool clear_ascii_hs_by_cleark = ((m_dsw_u39->read()&0x003) == 0x001);
+		if(clear_ascii_hs_by_cleark)
+		{
+			m_write_enql_cb(0);
+			m_write_ackl_cb(0);
+			m_write_nakl_cb(0);
+		}
+	}
+}
+
+void tandberg_tdv2100_disp_logic_device::w_linek(int state)
+{
+	if(state)
+	{
+		bool force_on_line = m_dsw_u39->read()&0x020;
+		if(force_on_line)
+		{
+			data_terminal_ready = true;
+		}
+		else
+		{
+			// Try to toggle RS232 DTR
+			if(rx_handshake)
+			{
+				data_terminal_ready = false;
+			}
+			else
+			{
+				data_terminal_ready = !data_terminal_ready;
+			}
+		}
+
+		if(!data_terminal_ready)
+		{
+			request_to_send = false;
+			m_rs232->write_rts(request_to_send);
+			bool auto_tx_handshake = m_sw_rs232_settings->read()&0x20;
+			if(auto_tx_handshake)
+			{
+				tx_handshake = request_to_send;
+			}
+		}
+		m_rs232->write_dtr(data_terminal_ready);
+
+		bool auto_rx_handshake = m_sw_rs232_settings->read()&0x10;
+		if(auto_rx_handshake)
+		{
+			rx_handshake = data_terminal_ready;
+		}
+		bool hold_line_key_for_rts = ((m_dsw_u73->read()&0x3) == 0x1);
+		if(data_terminal_ready && hold_line_key_for_rts)
+		{
+			w_transk(1);
+		}
+		update_rs232_lamps();
+	}
+}
+
+void tandberg_tdv2100_disp_logic_device::w_transk(int state)
+{
+	if(state)
```

Comment:
Same here, you should be prepared for consecutive calls with the same value.

Reply:
Will fix